File: app/schemas/onboarding.py
"""
Pydantic schemas for the onboarding flow.
"""
from typing import Dict, List, Optional

from pydantic import BaseModel, Field, field_validator

from app.schemas.room import RoomCategoryCreate
from app.services.timezones import normalize_timezone


class OwnerPayload(BaseModel):
    name: str = Field(..., min_length=1, max_length=150)
    email: str = Field(..., min_length=3, max_length=200)
    phone: Optional[str] = Field(default=None, max_length=80)
    role: Optional[str] = Field(default=None, max_length=120)


class HotelIdentityPayload(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    timezone: str = Field(..., min_length=3, max_length=100)
    currency: str = Field(..., min_length=3, max_length=3)
    languages: List[str] = Field(default_factory=list, min_length=1)
    jurisdiction_code: str = Field(..., min_length=2, max_length=3)

    @field_validator("currency")
    @classmethod
    def normalize_currency(cls, value: str) -> str:
        return value.strip().upper()

    @field_validator("timezone")
    @classmethod
    def normalize_timezone_value(cls, value: str) -> str:
        return normalize_timezone(value)

    @field_validator("jurisdiction_code")
    @classmethod
    def normalize_jurisdiction(cls, value: str) -> str:
        return value.strip().upper()

    @field_validator("languages")
    @classmethod
    def normalize_languages(cls, value: List[str]) -> List[str]:
        cleaned = [item.strip() for item in value if item and item.strip()]
        if not cleaned:
            raise ValueError("At least one language is required")
        return cleaned


class DepositPolicyPayload(BaseModel):
    deposit_percentage: float = Field(..., ge=0, le=100)
    free_cancellation_hours: int = Field(..., ge=0, le=168)
    cancellation_penalty_percentage: float = Field(..., ge=0, le=100)


class ProviderSetupPayload(BaseModel):
    enabled: bool = False
    credentials: Dict[str, str] = Field(default_factory=dict)


class PaymentMethodsPayload(BaseModel):
    mercado_pago: ProviderSetupPayload = Field(default_factory=ProviderSetupPayload)
    paypal: ProviderSetupPayload = Field(default_factory=ProviderSetupPayload)
    stripe: ProviderSetupPayload = Field(default_factory=ProviderSetupPayload)


class OTAChannelsPayload(BaseModel):
    booking: ProviderSetupPayload = Field(default_factory=ProviderSetupPayload)
    expedia: ProviderSetupPayload = Field(default_factory=ProviderSetupPayload)
    despegar: ProviderSetupPayload = Field(default_factory=ProviderSetupPayload)


class SubscriptionChoicePayload(BaseModel):
    plan_code: str = Field(..., min_length=3, max_length=20)
    start_trial: bool = False

    @field_validator("plan_code")
    @classmethod
    def normalize_plan_code(cls, value: str) -> str:
        return value.strip().lower()


class CategoriesPayload(BaseModel):
    categories: List[RoomCategoryCreate]


class RoomInput(BaseModel):
    room_number: str = Field(..., min_length=1, max_length=10)
    floor: int = Field(ge=0)
    category_code: str = Field(..., min_length=1, max_length=20)


class RoomsPayload(BaseModel):
    rooms: List[RoomInput]


class StaffMember(BaseModel):
    name: str = Field(..., min_length=1, max_length=150)
    role: Optional[str] = Field(default=None, max_length=120)
    email: Optional[str] = Field(default=None, max_length=200)
    phone: Optional[str] = Field(default=None, max_length=80)


class StaffPayload(BaseModel):
    staff: List[StaffMember]


class OnboardingStatus(BaseModel):
    hotel_id: int
    completed: bool
    steps: Dict[str, bool]
    missing_steps: List[str]
    gates: Optional[dict] = None
    counts: Dict[str, int]
    owner: Optional[dict] = None
    hotel_identity: Optional[dict] = None
    deposit_policy: Optional[dict] = None
    payment_methods: Optional[dict] = None
    ota_channels: Optional[dict] = None
    subscription_choice: Optional[dict] = None
    current_subscription: Optional[dict] = None
    categories: List[dict] = Field(default_factory=list)
    rooms: List[dict] = Field(default_factory=list)
    staff: List[dict] = Field(default_factory=list)

File: frontend/src/api/onboarding.ts
import { apiFetch, type SessionLike } from "./client";

export type OnboardingProviderSetup = {
  enabled: boolean;
  credentials?: Record<string, string>;
  has_credentials?: boolean;
  credential_fields?: string[];
};

export type OnboardingStatus = {
  hotel_id: number;
  completed: boolean;
  steps: Record<string, boolean>;
  missing_steps: string[];
  gates?: {
    can_finish: boolean;
    missing: string[];
  } | null;
  counts: Record<string, number>;
  owner?: Record<string, unknown> | null;
  hotel_identity?: Record<string, unknown> | null;
  deposit_policy?: Record<string, unknown> | null;
  payment_methods?: Record<string, OnboardingProviderSetup> | null;
  ota_channels?: Record<string, OnboardingProviderSetup> | null;
  subscription_choice?: Record<string, unknown> | null;
  current_subscription?: Record<string, unknown> | null;
  categories?: CategoryPayload[];
  rooms?: RoomPayload[];
  staff?: StaffPayload[];
};

export type OwnerPayload = {
  name: string;
  email: string;
  phone?: string | null;
  role?: string | null;
};

export type HotelIdentityPayload = {
  name: string;
  timezone: string;
  currency: string;
  languages: string[];
  jurisdiction_code: string;
};

export type DepositPolicyPayload = {
  deposit_percentage: number;
  free_cancellation_hours: number;
  cancellation_penalty_percentage: number;
};

export type CategoryPayload = {
  name: string;
  code: string;
  description?: string;
  base_price_per_night: number;
  max_occupancy: number;
  amenities?: string;
};

export type RoomPayload = {
  room_number: string;
  floor: number;
  category_code: string;
};

export type StaffPayload = {
  name: string;
  role?: string;
  email?: string;
  phone?: string;
};

export type PaymentMethodsPayload = {
  mercado_pago: OnboardingProviderSetup;
  paypal: OnboardingProviderSetup;
  stripe: OnboardingProviderSetup;
};

export type OTAChannelsPayload = {
  booking: OnboardingProviderSetup;
  expedia: OnboardingProviderSetup;
  despegar: OnboardingProviderSetup;
};

export type SubscriptionChoicePayload = {
  plan_code: string;
  start_trial: boolean;
};

export const getOnboardingStatus = (session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/status", { session });

export const setOwner = (payload: OwnerPayload, session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/owner", { method: "POST", data: payload, session });

export const setHotelIdentity = (payload: HotelIdentityPayload, session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/identity", { method: "POST", data: payload, session });

export const setCategories = (categories: CategoryPayload[], session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/categories", {
    method: "POST",
    data: { categories },
    session
  });

export const setRooms = (rooms: RoomPayload[], session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/rooms", {
    method: "POST",
    data: { rooms },
    session
  });

export const setDepositPolicy = (payload: DepositPolicyPayload, session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/policy", {
    method: "POST",
    data: payload,
    session
  });

export const setPaymentMethods = (payload: PaymentMethodsPayload, session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/payments", {
    method: "POST",
    data: payload,
    session
  });

export const setOtaChannels = (payload: OTAChannelsPayload, session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/ota", {
    method: "POST",
    data: payload,
    session
  });

export const setSubscriptionChoice = (payload: SubscriptionChoicePayload, session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/subscription-choice", {
    method: "POST",
    data: payload,
    session
  });

export const setStaff = (staff: StaffPayload[], session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/staff", {
    method: "POST",
    data: { staff },
    session
  });

export const finishOnboarding = (session?: SessionLike) =>
  apiFetch<OnboardingStatus>("/api/onboarding/finish", {
    method: "POST",
    session
  });
